File: apps/api/openapi_server/dbmodels/challenge_platform.py
import datetime
from bson import ObjectId
from mongoengine import (
    DateTimeField,
    Document,
    ObjectIdField,
    StringField,
    URLField,
    IntField,
)  # noqa: E501


class ChallengePlatform(Document):
    id = ObjectIdField(primary_key=True, default=ObjectId)
    name = StringField(required=True, unique=True)
    displayName = StringField(required=True, unique=True)
    websiteUrl = URLField(required=True)
    avatarUrl = URLField()
    createdAt = DateTimeField(required=True, default=datetime.datetime.now)
    updatedAt = DateTimeField(required=True, default=datetime.datetime.now)
    v = IntField(db_field="__v")

    def to_dict(self):
        doc = self.to_mongo().to_dict()
        doc["id"] = str(self.pk)
        return doc

File: apps/challenge-db-cli/src/models/challenge-platform.ts
import { model, ObjectId, Schema } from 'mongoose';
import validator from 'validator';

export interface ChallengePlatform {
  _id: ObjectId;
  name: string;
  displayName: string;
  websiteUrl?: string;
  avatarUrl?: string;
  createdAt: Date;
  updatedAt: Date;
}

const options = {
  collection: 'challenge_platform',
  timestamps: true,
};

export const ChallengePlatformSchema = new Schema<ChallengePlatform>(
  {
    _id: { type: Schema.Types.ObjectId, required: true },
    name: { type: String, required: true, unique: true },
    displayName: { type: String, required: true, unique: true },
    websiteUrl: {
      type: String,
      validate: [validator.isURL, 'invalid websiteUrl'],
    },
    avatarUrl: {
      type: String,
      validate: [validator.isURL, 'invalid avatarUrl'],
    },
  },
  options
);

export const ChallengePlatformModel = model(
  'ChallengePlatform',
  ChallengePlatformSchema
);
